ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -234,12 +198,16 @@ impl<C: ShardedContext> ShuffleContext for C {}
 ///
 /// [`ShuffleShare`] and [`Shuffleable`] are added to bridge the gap. They can be implemented for
 /// arbitrary structs as long as `Add` operation can be defined on them.
-pub trait ShuffleShare: Sendable + Clone + FromRandom + Add<Output = Self> {}
+pub trait ShuffleShare: BooleanArray + Serializable + FromRandom {}
 
-impl<V: Sendable + Clone + FromRandom + Add<Output = Self>> ShuffleShare for V {}
+impl<V: BooleanArray + Serializable + FromRandom> ShuffleShare for V {}
 
 /// Trait for shuffle inputs that consists of two values (left and right).
-pub trait Shuffleable: Send + 'static {
+// The `From` and `Into` bounds are necessary to work with routines that have not been
+// updated to use the `Shuffleable` trait.
+pub trait Shuffleable:
+    From<AdditiveShare<Self::Share>> + Into<AdditiveShare<Self::Share>> + Send + 'static
```

Comment:
The `From` and `Into` bounds are necessary to interoperate with the malicious shuffle routines that don't understand `Shuffleable` yet. I have a future change that updates things to use `Shuffleable` and gets rid of these bounds.

(At the moment, `Shuffleable` is only implemented for `AdditiveShare<V: BooleanArray>`, so these bounds are trivially satisfied. But we want to implement `Shuffleable` for other types.)